!!! 9.0: Serious root nodes

nezaniel opened the pull request:
This cleans up the handling of root nodes and absolute node paths.

Root nodes don't have names but are to be resolved by type (if not identifier).
Thus, the following adjustments have been made:

* Uniqueness of node type name for root nodes is now enforced on the write side (threw an exception on the read side until now)
* The ContentSubgraphInterface has a new method "findRootNodeByType" similar to the ContentGraphInterface
* Node paths can now be absolute by having a root node type name defined. Absolute paths can be serialized as '/<Neos.ContentRepository.Root>/my/path' with the first path segment being the root node type name
* Tests for node paths have been written
* ContentSubgraphInterface::findNodeByPath has been adjusted to accept either the path's root node aggregate type or the starting node aggregate ID
* ContentSubgraphInterface::retrieveNodePath has been adjusted to return absolute paths with root node aggregate type https://github.com/neos/neos-development-collection/issues/4323
* ContentSubgraphInterface::retrieveNodePath also has been adjusted to fill up "missing" node names with node aggregate IDs https://github.com/neos/neos-development-collection/issues/4331
* Adjusted the Neos endpoint API to support serialized absolute node paths for the reference editor 

**Review instructions**

Besides the tests, try out a reference editor starting point like '/<Neos.ContentRepository.Root>/neosdemo'

**Checklist**

- [x] Code follows the PSR-12 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the 9.0 branch

Commit:
Enforce type-uniqueness of root nodes

Commit:
Remove usage of dynamic properties in tests

Commit:
Implement ContentSubgraphInterface::findRootNodeByType

Commit:
Refactor node paths to being able to be absolute

mhsdesign commented:
Should we provide a migration for people using „sites“ and rewrite this to Neos.Neos:Sites?

nezaniel commented:
Sure

> Should we provide a migration for people using „sites“ and rewrite this to Neos.Neos:Sites?

Sure thing, feel free to do it :)

Commit:
Add test cases for ContentSubgraphInterface::findNodeByPath

Commit:
Adjust ContentSubgraph to retrieve node paths including root node aggregate type

also fill missing segments with node aggregate IDs

nezaniel commented:
Still missing: UI integration, will follow in a bit

Commit:
Adjust test cases to use distinct root node types

Commit:
Properly serialize adjusted node paths

Commit:
Pacify linter

Commit:
Adjust Neos nodes endpoint to paths

Commit:
Pacify linter

Commit:
Adjust documentation about paths

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodePath.php:
```diff
@@ -90,19 +116,27 @@ public function getDepth(): int
+            && $this->rootNodeTypeName?->value === $other->rootNodeTypeName?->value;
+    }
+
+    public function __toString(): string
+    {
+        return $this->rootNodeTypeName
+            ? rtrim('/<' . $this->rootNodeTypeName->value . '>/' . (ltrim($this->path, '/')), '/')
+            : $this->path;
     }
```
i thought we wanted to ban `__toString` from value objects? see @bwaidelich 's #4156

bwaidelich replied:
I agree, we should never rely on the actual result of `__toString()` and therefore it should not be there to avoid errors.

With a dedicated `AbsoluteNodePath` VO this could be solved I suppose

nezaniel replied:
you're right, what I wanted to do was toString(), not __toString() /facepalm
(alternatively something like serializeToString() or stringSerialize()

nezaniel replied:
has been adjusted, we still need string serialization for absolute node paths, though

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/ContentSubgraphInterface.php:
```diff
@@ -141,12 +151,15 @@ public function findBackReferences(NodeAggregateId $nodeAggregateId, Filter\Find
 
     /**
-     * Find a single node underneath $startingNodeAggregateId that matches the specified $path
+     * Find a single node underneath
+     * - the given root node type name if defined in the path
+     * - the given starting node aggregate ID else
+     * that matches the specified $path
      *
      * NOTE: This operation is most likely to be deprecated since the concept of node paths is not really used in the core, and it has some logical issues
      * @return Node|null the node that matches the given $path, or NULL if no node on that path is accessible
      */
-    public function findNodeByPath(NodePath $path, NodeAggregateId $startingNodeAggregateId): ?Node;
+    public function findNodeByPath(NodePath $path, ?NodeAggregateId $startingNodeAggregateId): ?Node;
```
im not sure i like this conditional behaviour cluttered into one method ...

Would it really make the api easier to use?

what do you think of this alternative, is it too complex?
```php
$node = $path->isAbsolute()
   ? $subgraph->findNodeByAbsolutePath($path)
   : $subgraph->findNodeByPath($path, $startingNodeAggregateId)
```

mficzel replied:
Also do not like the second argument that can be null depending on the first one that much. For me it would be fine to always require a nodeAggregateId.

The alternative would be to have a distinct AbsoluteRootPath value object wich is a bit overdone. Will test this afternoon.

bwaidelich replied:
+1 for two distinct methods and +.9 for two dedicated VOs representing absolute or relative paths

nezaniel replied:
yes, the more I think about it the more I prefer separate VOs and separate methods.

nezaniel replied:
Say hello to AbsoluteNodePath and ContentSubgraphInterface::findNodeByAbsolutePath

mhsdesign reviewed (commented):
Just some early thoughts ;)

Commit:
Separate absolute from relative paths

Commit:
Adjust path tests

Commit:
Properly render node helper path exception

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/ContentSubgraphInterface.php:
```diff
@@ -148,13 +158,21 @@ public function countBackReferences(NodeAggregateId $nodeAggregateId, Filter\Cou
      * @throws \InvalidArgumentException if the node path could not be retrieved because it is inaccessible or contains no valid path. The latter can happen if any node in the hierarchy has no name
      */
-    public function retrieveNodePath(NodeAggregateId $nodeAggregateId): NodePath;
+    public function retrieveNodePath(NodeAggregateId $nodeAggregateId): AbsoluteNodePath;
```
I like this change. But i think we need to communicate it as breaking, as it IS breaking. The Neos UI for example wont survive that without adjustments ^^

~and there are a few more usages of  `retrieveNodePath` which need to be checked.~
Edit: I think you found them all ^^ in the dev collection

nezaniel replied:
I guess the Neos UI must be adjusted as well, yes. But I don't think there is a better way to solve this

mhsdesign replied:
Yes i agree, thanks for https://github.com/neos/neos-ui/pull/3530

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/AbsoluteNodePath.php:
```diff
@@ -0,0 +1,111 @@
+        );
+    }
+
+    public function isRoot(): bool
+    {
+        return $this->path->value === '/';
+    }
```
I think `isRoot` in combination with `$rootNodeTypeName` is a bit unlucky and misunderstandable

nezaniel replied:
Yep, I mixed this up several times myself, but semantically this is correct. All absolute paths must _have_ a root, but not all of them _are_ the root path

mhsdesign replied:
Okay then lets put exactly this into the doc comment ^^

nezaniel replied:
done

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/AbsoluteNodePath.php:
```diff
@@ -0,0 +1,111 @@
+        return $this->path->getParts();
+    }
+
+    public function getDepth(): int
+    {
+        return count($this->path->getParts());
+    }
```
shouldn't the depth include the $rootNodeTypeName? so +1

nezaniel replied:
imho the root node should have depth 0, so nope ^^

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodePath.php:
```diff
@@ -77,32 +81,37 @@ public function appendPathSegment(NodeName $nodeName): self
-                'Depth of relative node path "%s" cannot be determined',
-                $this->value
-            ), 1548162166);
-        }
-        return count($this->getParts());
+        throw new \RuntimeException(sprintf(
+            'Depth of relative node path "%s" cannot be determined',
+            $this->value
+        ), 1548162166);
     }
```
makes sense, but we might aswell remove it?

nezaniel replied:
yep, simply forgot about it; It's dead code und must be gone. It's kind of breaking though, see below

nezaniel replied:
done

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodePath.php:
```diff
@@ -63,11 +72,6 @@ public function isRoot(): bool
         return $this->value === '/';
     }
 
-    public function isAbsolute(): bool
-    {
-        return strpos($this->value, '/') === 0;
-    }
-
```
makes sense that this will be remove, commenting on it as it _was_ api. But useless api^^

Commit:
remove obsolete node path methods

Commit:
Describe AbsoluteNodePath::isRoot

mhsdesign reviewed (approved):
Thank you so much for the adjustments. With the change in the UI https://github.com/neos/neos-ui/pull/3530 it should work fine.

So dare i say it: Fine by reading ^^ (count it as a 3/4 review for now, ill see if i find the time to experiment with it.)

mficzel commented:
Almost there and a huge improvement ... i am still looking into some issues when the references editor has nodes from foreign roots selected.

mficzel reviewed (changes requested):
If nodes were selected before the endpoint `neos/service/nodes` gets called with the ids in
parameter `nodeIdentifiers` and not in `nodeIds` as the \Neos\Neos\Controller\Service\NodesController::indexAction expects. 

Other than that this works like a charm even with taxonomies in a separate root.

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php:
```diff
@@ -250,7 +282,45 @@ public function retrieveNodePath(NodeAggregateId $nodeAggregateId): NodePath
+        $pathSegments = [];
+        foreach ($rows as $i => $row) {
+            if ($i !== 0) { // skip the root node, it's resolved by node type name
+                if ($row['name']) {
```
~Isn’t here an isset missing?~
Edit: no it will be null

also the retrieve operation now seems more expensive, but I think that’s fine, since it should not be used that much

mhsdesign replied:
@nezaniel i would find it actually cool to use findAncestors https://github.com/neos/neos-development-collection/pull/4349 instead and just postprocess the nodes.
This would would reduce complexity quite a bit ^^

bwaidelich replied:
Wow, this is the most complex code in this class now and it exceeds my mental capabilities

> This would would reduce complexity quite a bit ^^

Yes, please :)

mhsdesign replied:
We said, we want to finish https://github.com/neos/neos-development-collection/pull/4349 first ^^

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php:
```diff
@@ -181,9 +199,23 @@ public function findParentNode(NodeAggregateId $childNodeAggregateId): ?Node
+            : null;
+    }
+
+    private function findNodeByPathFromStartingNode(Node $currentNode, NodePath $path): ?Node
```
Do we need to adjust this method to work with ~absolute~ paths, where there is a filled up node name as node id?

Maybe we should consider making the Nodepath a collection of NodeName|NodeAggId

Edit: i meant normal nodePath's

mhsdesign replied:
I actually dislike to fill up a nodepath with node ids. It totally overqualifies the result.

How about defining an absolute node path like:

if there are all named nodes:

`/<Neos.Neos.Sites>/my-site/bla/blie`

but if `bla` has no node name, it currently would insert its id:

`/<Neos.Neos.Sites>/my-site/1234567543256746578/blie`

but to identify the node `blie` at the end, the last part is already enough. So how about we use the following notation? @nezaniel @mficzel ?

`/1234567543256746578/blie`


or maybe with a different syntax:

`/<1234567543256746578>/blie`

bwaidelich replied:
I find it a bit confusing that we now have `findNodeByPath` and `findNodeByPathFromStartingNode`, both basically doing the same but with the arguments swapped.
But part of the confusion is maybe because this internal method is placed between the API methods. I would suggest to move it out of the way, down to the other private method.

> Do we need to adjust this method to work with absolute path, where there is a filled up node name as node id?

@mhsdesign This is a private method, why would we need to accept absolute paths?

> Maybe we should consider making the Nodepath a collection of NodeName|NodeAggId

Maybe it makes sense to solve this on the lowest level really, i.e. by changing the `Node` constructor from:
```php
public function __construct(
    // ...
    public readonly NodeAggregateId $nodeAggregateId,
    // ...
    public readonly ?NodeName $nodeName,
) {
}
```
to s.th. like:
```php
public NodeName $name;

public function __construct(
    // ...
    public readonly NodeAggregateId $nodeAggregateId,
    // ...
    ?NodeName $nodeName,
) {
    $this->name = $nodeName ?? NodeName::fromString($nodeAggregateId->value);
}
```

mhsdesign replied:
@bwaidelich sorry my bad, i meant 

Do we need to adjust this method to work with paths, where there is a filled up node name as node id?

I actually still like my proposal `/<1234567543256746578>/blie`

mficzel replied:
@mhsdesign `/1234567543256746578/blie` would mean to extend the concept of absolute pathes further and introduce new features. Would not recommend especially as we want to limit the use to specific cases.

One thing to discuss is removing the fallback to aggregateIds so pathes would only work on root and named nodes.

Commit:
Fall back to legacy $nodeIdentifiers in NodesController

nezaniel commented:
> If nodes were selected before the endpoint `neos/service/nodes` gets called with the ids in parameter `nodeIdentifiers` and not in `nodeIds` as the \Neos\Neos\Controller\Service\NodesController::indexAction expects.

I've added a fallback to the legacy variable, please check whether this works out

mhsdesign reviewed (changes requested):
see my comments

https://github.com/neos/neos-development-collection/pull/4339#discussion_r1235965808
https://github.com/neos/neos-development-collection/pull/4339#discussion_r1236419042

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/AbsoluteNodePath.php:
```diff
@@ -0,0 +1,115 @@
+
+    public static function fromString(string $value): self
+    {
+        if (\str_starts_with($value, '/<')) {
+            $pivot = \mb_strpos($value, '>');
+            if ($pivot > 0) {
+                $nodeTypeName = NodeTypeName::fromString(\mb_substr($value, 2, $pivot - 2));
+                $path = \mb_substr($value, $pivot + 2) ?: '/';
+            } else {
+                throw AbsoluteNodePathIsInvalid::becauseItDoesNotMatchTheRequiredPattern($value);
+            }
+        } else {
+            throw AbsoluteNodePathIsInvalid::becauseItDoesNotMatchTheRequiredPattern($value);
+        }
```
Just a cosmetic nitpick, but an early return would save some cognitive load:
```suggestion
        if (!\str_starts_with($value, '/<')) {
            throw AbsoluteNodePathIsInvalid::becauseItDoesNotMatchTheRequiredPattern($value);
        }
        $pivot = \mb_strpos($value, '>');
        if ($pivot === false) {
            throw AbsoluteNodePathIsInvalid::becauseItDoesNotMatchTheRequiredPattern($value);
        }
        $nodeTypeName = NodeTypeName::fromString(\mb_substr($value, 2, $pivot - 2));
        $path = \mb_substr($value, $pivot + 2) ?: '/';
```

mhsdesign replied:
i plan to refactor this to regex either way ^^

in the same run as fixing the last remaining unstrict dtos: https://github.com/neos/neos-development-collection/issues/4133

bwaidelich commented on Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/ContentSubhypergraph.php:
```diff
@@ -335,9 +364,9 @@ private function findAnySiblings(
+    public function retrieveNodePath(NodeAggregateId $nodeAggregateId): AbsoluteNodePath
     {
-        return NodePath::fromString('/');
+        return AbsoluteNodePath::fromString('/');
```
This should fail because it does not satisfy the `AbsoluteNotePath` pattern

bwaidelich commented on Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/ContentSubhypergraph.php:
```diff
@@ -235,24 +255,33 @@ public function findParentNode(NodeAggregateId $childNodeAggregateId): ?Node
+            : null;
+    }
+
+    private function findNodeByPathFromStartingNode(Node $currentNode, NodePath $path): ?Node
```
Same as above, I would suggest to move private methods together (and consider swapping the arguments)

bwaidelich commented on Neos.ContentRepository.Core/Classes/SharedModel/Exception/RootNodeAggregateDoesNotExist.php:
```diff
@@ -0,0 +1,21 @@
+/**
+ * @api Userland code might have to react to this
+ */
+final class RootNodeAggregateDoesNotExist extends \DomainException
```
same here

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraph.php:
```diff
@@ -141,7 +145,7 @@ public function findRootNodeAggregateByType(
 
         if ($rootNodeAggregate === null) {
-            throw new \RuntimeException('Root Node Aggregate not found');
+            throw RootNodeAggregateDoesNotExist::butWasExpectedTo($nodeTypeName);
```
sweet, but this is a breaking change because the exception base type changed, see comment below.
But apart from that I wonder if this should actually changed to return a nullable like most (all?) other `find*()` methods – Alternatively, this should be renamed to s.th. like `getRootNodeAggregateByType()`

mhsdesign replied:
nullable seems to be more consistent. And there is no guarantee that a root node of say, sitegeist taxonomies exists so yes

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:
```diff
@@ -139,6 +141,22 @@ protected function requireNodeTypeToNotBeOfTypeRoot(NodeType $nodeType): void
+        ContentStreamId $contentStreamId,
+        ContentRepository $contentRepository
+    ): void {
+        try {
+            $rootNodeAggregate = $contentRepository->getContentGraph()->findRootNodeAggregateByType(
+                $contentStreamId,
+                $nodeTypeName
+            );
+            throw RootNodeAggregateTypeIsAlreadyOccupied::butWasExpectedNotTo($nodeTypeName);
+        } catch (RootNodeAggregateDoesNotExist $exception) {
+            // all is well
+        }
```
empty `catch` blocks always make me suspicious (and in general: using exceptions for control flow).
With the change suggested above, this could be changed to:

```suggestion
        $rootNodeAggregate = $contentRepository->getContentGraph()->findRootNodeAggregateByType(
            $contentStreamId,
            $nodeTypeName
        );
        if ($rootNodeAggregate !== null) {
            throw RootNodeAggregateTypeIsAlreadyOccupied::butWasExpectedNotTo($nodeTypeName);
        }
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/AbsoluteNodePath.php:
```diff
@@ -0,0 +1,115 @@
+{
+    private function __construct(
+        public readonly NodeTypeName $rootNodeTypeName,
+        public readonly NodePath $path
```
for `NodePath` it's called `$value` and I think it could be named the same here for consistency (even though there is a root node type name attached as well)

grebaldi replied:
I'd opt for a more precise `$relativePathAfterRoot`. `$value` I'd find confusing...

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/AbsoluteNodePath.php:
```diff
@@ -0,0 +1,115 @@
+    ) {
+    }
+
+    public static function fromComponents(
```
Not that important but I suggest to make this more explicit even:
```suggestion
    public static function fromRootNodeTypeNameAndPath(
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/SharedModel/Exception/AbsoluteNodePathIsInvalid.php:
```diff
@@ -0,0 +1,32 @@
+ *
+ * @api because this might have to be handled in the application layer
+ */
+final class AbsoluteNodePathIsInvalid extends \DomainException
```
The [explanation for DomainExceptions on php.net](https://www.php.net/manual/en/class.domainexception.php) is

> Exception thrown if a value does not adhere to a defined valid data domain.

or, in German:

> Ausnahme, die ausgelöst wird, wenn ein Wert nicht in einem definierten gültigen Datenbereich liegt.

I don't think that this is the correct type here as "domain" does not refer to DDD domain..

I would suggest to extend `RuntimeException` or `InvalidArgumentException`

mhsdesign replied:
i actually prefer `\DomainException` for these kind of exceptions, but only because i use them all the time for it ^^

i preserve `RuntimeException` for something more dynamic and unlikely. And `InvalidArgumentException` for cases in php7.4 where union types didnt exist or a string must not be empty in some random method ^^

But this is just preference and i have no data/facts to back this up.

grebaldi replied:
Well, I think that exceptions of the `*IsInvalid` kind (as long as validity is in regard to a specific data format) meet precisely the definition of 

> [...] a value does not adhere to a defined valid data domain

Therefore, I'd argue that `\DomainException` is actually quite correct in this instance.

Furthermore, I would argue that `RuntimeException` doesn't fit this case. It is defined as:

> Exception thrown if an error which can only be found on runtime occurs.

...which is not true in this case (the format of an absolute node path does not depend on any runtime conditions).

Likewise, `InvalidArgumentException` doesn't fit either. It is defined as:

> Exception thrown if an argument is not of the expected type.

...but this exception occurs when somebody attempts to create an absolute node path, which doesn't necessarily occur in the context of that path being an argument to anything.

That being said, I'm generally more in favor of simply extending `\Exception`, to avoid such debate :laughing:

bwaidelich commented on Neos.ContentRepository.Core/Classes/SharedModel/Exception/RootNodeAggregateTypeIsAlreadyOccupied.php:
```diff
@@ -0,0 +1,21 @@
+/**
+ * @api Userland code might have to react to this
+ */
+final class RootNodeAggregateTypeIsAlreadyOccupied extends \DomainException
```
and here

grebaldi replied:
```suggestion
final class RootNodeAggregateTypeIsAlreadyInUse extends \RuntimeException
```

Two changes:

- `*Occupied` -> `*InUse` (as per my earlier suggestion to change the naming slightly to better match the semantics)
- `\DomainException` -> `\RuntimeException` (funny, here I think `\DomainException` is indeed the wrong base class, since the root node type being already in use is a condition that can only happen at runtime)

Beyond this, I have a question:

Why is this exception speaking of a `RootNodeAggregateType` instead of a `RootNodeType`? Are (root) node aggregate types a concept that exists?

bwaidelich commented on Neos.ContentRepository.Core/Tests/Behavior/Features/Bootstrap/EventSourcedTrait.php:
```diff
@@ -117,6 +118,8 @@ trait EventSourcedTrait
     private ContentRepository $contentRepository;
     private ContentRepositoryInternals $contentRepositoryInternals;
 
+    private ?UserId $currentUserId = null;
```
What is this about?

bwaidelich commented on Neos.ContentRepository.Security/Classes/Authorization/Privilege/Node/NodePrivilegeContext.php:
```diff
@@ -61,7 +61,7 @@ public function isAncestorNodeOf(string $nodePathOrIdentifier): bool
         }
 
-        return str_starts_with($nodePathOrResult, $this->getSubgraph()->retrieveNodePath($this->node->nodeAggregateId)->value);
+        return str_starts_with($nodePathOrResult, $this->getSubgraph()->retrieveNodePath($this->node->nodeAggregateId)->serializeToString());
```
This would require `$nodePathOrIdentifier` to have a syntax of `<NodeType>/foo/bar` – is that really the case? If so, the doc comments should be adjusted accordingly

bwaidelich commented on Neos.Neos/Classes/Fusion/Helper/NodeHelper.php:
```diff
@@ -53,10 +54,13 @@ public function nearestContentCollection(Node $node, string $nodePath): Node
-                $node,
-                NodePath::fromString($nodePath)
-            );
+            $nodePath = AbsoluteNodePath::tryFromString($nodePath)
+                ?: NodePath::fromString($nodePath);
```
This is quite implicit and uses exceptions for control flow. What about something like:
```suggestion
            $nodePath = AbsoluteNodePath::isAbsolutePath($nodePath) ? AbsoluteNodePath::fromString($nodePath) : NodePath::fromString($nodePath);
```
(or maybe a better method name than `isAbsolutePath()`).

Apart from that, static analysis should complain about changing the type of `$nodePath` from `string` to `AbsoluteNodePath|NodePath` – it should get a new variable name.

And while we're on it, we could remove the whole `else` block since there is an early return in place..

In general, this method is quite overcomplex and could be simplified to
```php
public function nearestContentCollection(Node $node, string $nodePath): Node
{
    $contentCollectionType = 'Neos.Neos:ContentCollection';
    if ($node->nodeType->isOfType($contentCollectionType)) {
        return $node;
    }
    if ($nodePath === '') {
        throw new Exception(sprintf(
            'No content collection of type %s could be found in the current node and no node path was provided.'
            . ' You might want to configure the nodePath property'
            . ' with a relative path to the content collection.',
            $contentCollectionType
        ), 1409300545);
    }
    $subgraph = $this->contentRepositoryRegistry->subgraphForNode($node);
    if (AbsoluteNodePath::isAbsolutePath($nodePath)) {
        $subNode = $subgraph->findNodeByAbsolutePath(AbsoluteNodePath::fromString($nodePath));
    } else {
        $subNode = $subgraph->findNodeByPath(NodePath::fromString($nodePath));
    }
    if ($subNode === null) {
        throw new Exception(...);
    }
    if (!$subNode->nodeType->isOfType($contentCollectionType)) {
        throw new Exception(...);
    }
    return $subNode;
}
```

But that makes me wonder about the API.. If you specify an absolute `$nodePath` the `$node` argument is completely ignored (unless it is a content collection itself).
This screams "instability" :)

bwaidelich commented on Neos.Neos/Classes/Controller/Service/NodesController.php:
```diff
@@ -107,19 +110,26 @@ class NodesController extends ActionController
-
-        $nodeAddress = $contextNode
-            ? NodeAddressFactory::create($contentRepository)->createFromUriString($contextNode)
+        $nodePath = $contextNode
+            ? AbsoluteNodePath::tryFromString($contextNode)
             : null;
+        $nodeAddress = null;
+        if (!$nodePath) {
+            $nodeAddress = $contextNode
+                ? NodeAddressFactory::create($contentRepository)->createFromUriString($contextNode)
+                : null;
+        }
```
I have a hard time trying to understand what this code does.. Couldn't this be simplified to s.th. like:
```suggestion
        $nodeAddress = null;
        if ($contextNode !== null && AbsoluteNodePath::isAbsolutePath($contextNode)) {
            $nodeAddress = NodeAddressFactory::create($contentRepository)->createFromUriString($contextNode);
        }
```

bwaidelich reviewed (commented):
Great direction.
A first batch of comments :)

mficzel commented:
I can confirm that this now works with taxonomies like a charm. If anyone wants to play with that https://github.com/sitegeist/Sitegeist.Taxonomy/pull/66 ... will do a more detailed review later

Commit:
Adjust internal findNodeByPath methods

Commit:
Refactor absolute node path validation

Commit:
Use AbsoluteNodePath pattern matching for deciding what kind of path to use

Commit:
Base RootNodeAggregateDoesNotExist on previously used \RuntimeException

grebaldi commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php:
```diff
@@ -181,17 +199,19 @@ public function findParentNode(NodeAggregateId $childNodeAggregateId): ?Node
-        }
-        return $currentNode;
+
+        return $currentNode
```
I'd like to suggest renaming `$currentNode` to `$startingNode` for consistency.

grebaldi commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php:
```diff
@@ -181,17 +199,19 @@ public function findParentNode(NodeAggregateId $childNodeAggregateId): ?Node
+        $currentNode = $this->findRootNodeByType($path->rootNodeTypeName);
+
+        return $currentNode
+            ? $this->findNodeByPathFromStartingNode($path->path, $currentNode)
```
```suggestion
        $startingNode = $this->findRootNodeByType($path->rootNodeTypeName);

        return $startingNode
            ? $this->findNodeByPathFromStartingNode($path->path, $startingNode)
```

grebaldi commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php:
```diff
@@ -357,6 +415,18 @@ public function jsonSerialize(): array
 
+    private function findNodeByPathFromStartingNode(NodePath $path, Node $currentNode): ?Node
+    {
+        foreach ($path->getParts() as $edgeName) {
```
```suggestion
    private function findNodeByPathFromStartingNode(NodePath $path, Node $startingNode): ?Node
    {
        $currentNode = $startingNode;

        foreach ($path->getParts() as $edgeName) {
```

Nitpicks :roll_eyes: ! I know, I know :sweat_smile:

grebaldi commented on Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/ContentSubhypergraph.php:
```diff
@@ -235,25 +255,22 @@ public function findParentNode(NodeAggregateId $childNodeAggregateId): ?Node
-    ): ?Node {
+    public function findNodeByPath(NodePath $path, NodeAggregateId $startingNodeAggregateId): ?Node
+    {
         $currentNode = $this->findNodeById($startingNodeAggregateId);
```
```suggestion
        $startingNode = $this->findNodeById($startingNodeAggregateId);
```

Oh yes, I'm gonna do all of it :D

grebaldi commented on Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/ContentSubhypergraph.php:
```diff
@@ -235,25 +255,22 @@ public function findParentNode(NodeAggregateId $childNodeAggregateId): ?Node
+        return $currentNode
+            ? $this->findNodeByPathFromStartingNode($path, $currentNode)
+            : null;
+    }
```
```suggestion
        return $startingNode
            ? $this->findNodeByPathFromStartingNode($path, $startingNode)
            : null;
    }
```

grebaldi commented on Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/ContentSubhypergraph.php:
```diff
@@ -235,25 +255,22 @@ public function findParentNode(NodeAggregateId $childNodeAggregateId): ?Node
+
+        return $currentNode
+            ? $this->findNodeByPathFromStartingNode($path->path, $currentNode)
+            : null;
```
```suggestion
        $startingNode = $this->findRootNodeByType($path->rootNodeTypeName);

        return $startingNode
            ? $this->findNodeByPathFromStartingNode($path->path, $startingNode)
            : null;
```

grebaldi commented on Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/ContentSubhypergraph.php:
```diff
@@ -335,9 +352,9 @@ private function findAnySiblings(
+    public function retrieveNodePath(NodeAggregateId $nodeAggregateId): AbsoluteNodePath
     {
-        return NodePath::fromString('/');
+        return AbsoluteNodePath::fromString('/<Neos.ContentRepository:Root>');
```
Why is `'/<Neos.ContentRepository:Root>'` hard-coded here?

mhsdesign replied:
Because the Postgres adapter is not ready yet, there are many more fake method implementations in there

grebaldi commented on Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/ContentSubhypergraph.php:
```diff
@@ -463,6 +480,18 @@ public function countNodes(): int
 
+    private function findNodeByPathFromStartingNode(NodePath $path, Node $currentNode): ?Node
+    {
+        foreach ($path->getParts() as $edgeName) {
```
```suggestion
    private function findNodeByPathFromStartingNode(NodePath $path, Node $startingNode): ?Node
    {
        $currentNode = $startingNode;

        foreach ($path->getParts() as $edgeName) {
```

grebaldi commented on Neos.ContentRepository.Core/Classes/Feature/RootNodeCreation/RootNodeHandling.php:
```diff
@@ -70,6 +70,11 @@ private function handleCreateRootNodeAggregateWithNode(
         $nodeType = $this->requireNodeType($command->nodeTypeName);
         $this->requireNodeTypeToNotBeAbstract($nodeType);
         $this->requireNodeTypeToBeOfTypeRoot($nodeType);
+        $this->requireRootNodeTypeToBeUnoccupied(
```
As far as I can tell `requireRootNodeTypeToBeUnoccupied` is lacking an abstract declaration within this trait (but this is true of other similar `require*`-methods here as well).

More importantly, I'd like to suggest an alternative naming, because I have my doubts regarding the `occupied`-semantics. (Just to spell out the question that's bugging me: Who or what is occupied with or by whom or what?)

My suggestion would be to call this method: `requireRootNodeTypeNotToBeAlreadyInUse`

mhsdesign replied:
I actually like the first one better ^^

grebaldi commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/AbsoluteNodePath.php:
```diff
@@ -0,0 +1,121 @@
+     */
+    public function isRoot(): bool
+    {
+        return $this->path->value === '/';
```
Imho this should be encapsuled by `NodePath`, so that this can be expressed as:

```suggestion
        return $this->path->isEmpty();
```

grebaldi commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/AbsoluteNodePath.php:
```diff
@@ -0,0 +1,121 @@
+
+    public function getDepth(): int
+    {
+        return count($this->path->getParts());
```
Likewise, this should imho be encapsuled by `NodePath`, so that this can be expressed as:

```suggestion
        return $this->path->getLength();
```

grebaldi commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/NodePath.php:
```diff
@@ -63,11 +63,6 @@ public function isRoot(): bool
         return $this->value === '/';
```
Hmm... this is the one that imho should be used for `AbsoluteNodePath->isRoot()`. Yet, `isRoot` is semantically confusing now, because de facto a `NodePath` can never represent a "root" in any sense here (or am I wrong?).

I would therefore suggest to rename this method to `isEmpty`, which would be semantically correct in any case.

grebaldi commented on Neos.ContentRepository.Core/Classes/SharedModel/Exception/RootNodeAggregateDoesNotExist.php:
```diff
@@ -0,0 +1,21 @@
+<?php
+
```
```suggestion

/*
 * This file is part of the Neos.ContentRepository package.
 *
 * (c) Contributors of the Neos Project - www.neos.io
 *
 * This package is Open Source Software. For the full copyright and license
 * information, please view the LICENSE file which was distributed with this
 * source code.
 */

```

grebaldi reviewed (commented):
Hi  @nezaniel,

I haven't tested anything yet, just went through the code. The result is mostly nitpicks :)

And I have to say: Reading through the new CR code is quite the delight :+1:

mhsdesign commented:
As we already touch a lot of code we might aswell consider putting the following cleanup in here:

NodePaths should imo contain a collection of NodeNames as array and have a serialise toString representation which is calculated.

nezaniel committed:
Update Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php

Co-authored-by: Wilhelm Behncke <2522299+grebaldi@users.noreply.github.com>

nezaniel committed:
Update Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php

Co-authored-by: Wilhelm Behncke <2522299+grebaldi@users.noreply.github.com>

nezaniel committed:
Update Neos.ContentRepository.Core/Classes/SharedModel/Exception/RootNodeAggregateDoesNotExist.php

Co-authored-by: Wilhelm Behncke <2522299+grebaldi@users.noreply.github.com>

nezaniel committed:
Update Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/ContentSubhypergraph.php

Co-authored-by: Wilhelm Behncke <2522299+grebaldi@users.noreply.github.com>

nezaniel committed:
Update Neos.ContentGraph.PostgreSQLAdapter/src/Domain/Repository/ContentSubhypergraph.php

Co-authored-by: Wilhelm Behncke <2522299+grebaldi@users.noreply.github.com>

Commit:
Rename currentNode to startingNode

Commit:
refactor (absolute) path logic

Commit:
Refactor node paths

Commit:
Merge branch '9.0' into seriousRootNodes

Commit:
refactor usage of ContentSubgraphInterface::retrieveNodePath to ::findAncestors

nezaniel commented:
I think I covered everything, if the tests pass and there are no more important concerns, I'd like to merge

nezaniel commented:
I think I covered everything, if the tests pass and there are no more important concerns, I'd like to merge

Commit:
move added fallback node identifiers to back of argument list

mhsdesign reviewed (approved):
Lets do it ;)

mficzel commented:
Just tested this and the last commit 61b4d4844f9b1f493a24e05b40bdf3a8e96e1871 introduces an error in the ui whenever a reference error is opened
`Could not convert target type "string": Warning: Array to string conversion in /tmp/neos/Development/SubContextddev/Cache/Code/Flow_Object_Classes/Neos_Flow_Property_TypeConverter_StringConverter.php line 114 - Check the logs for details` 

With the commit before 5b7e06c5f029f3a3a9910f638ea641f98e2f2842 i could verify:
- referencing of nodes in the same site (+ rendering links)
- linking of nodes nodes in other sites (+ rendering links)
- linking to nodes in totally different root-nodes (taxonomies)

mficzel committed:
BUGFIX: Use same order for parameter comments as is used in function

otherwise flow will get confused and throw exceptions that type conversion failed

mficzel commented:
Node: I could fix the problem with adjusting bringing the parameter comments into the same order as the function parameters. Pushed this change to the branch.

mficzel reviewed (approved):
Fine by testing .... what i tested: 

- referencing documents with configured starting point in the same site
- referencing documents with configured starting point in another site
- referencing other nodes (taxonomies) with a startingpoint other than `/<Neos.Neos:Sites>`

All three cases above worked and the nodes that were chosen by references editor could be used for traversal and link generation.

What i could not test is linking nodes in another cr which is uncharted territory even with this pr !!!

I personally cannot say that much about coding principles in the new cr other than that the code makes sense to me. So maybe count this as half an approval.

mficzel reviewed (changes requested):
Noticed a linting error in ci that obviously should be addressed. Seems to originate from changes made elsewhere since it also appeared in the last commit after running the test again.

Commit:
accept both single and multiple node IDs in NodesController

Commit:
Add missing type annotations

nezaniel commented:
I don't quite understand the linting error and since I cannot reproduce it locally, I'd just ignore it.
As for the test failure, I don't understand it either, and it should be resolved where it was introduced.

Go?

bwaidelich commented:
Thanks a lot for this.
~Unfortunately this seems to break Neos and https://github.com/neos/neos-ui/pull/3530 does not seem to cover everything~
Nevermind, I think it _does_ fix the issue :)

bwaidelich commented:
e.g. I immediately get
> Warning: Undefined property: Neos\ContentRepository\Core\Projection\ContentGraph\AbsoluteNodePath::$value in Cache/Code/Flow_Object_Classes/Neos_Neos_Ui_FlowQueryOperations_NeosUiDefaultNodesOperation.php line 88
